File: src/models/MaintenanceLog.js
const mongoose = require('mongoose');

const maintenanceLogSchema = new mongoose.Schema({
  vehicleId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Vehicle',
    required: [true, 'Le véhicule est requis'],
    index: true
  },
  label: {
    type: String,
    required: [true, 'Le libellé est requis'],
    trim: true,
    maxlength: [200, 'Le libellé ne peut pas dépasser 200 caractères']
  },
  category: {
    type: String,
    required: [true, 'La catégorie est requise'],
    enum: {
      values: [
        'vidange',
        'filtre_huile',
        'filtre_air',
        'filtre_essence',
        'bougies',
        'freins',
        'pneus',
        'batterie',
        'chaines',
        'liquide_refroidissement',
        'liquide_freins',
        'revision',
        'autre'
      ],
      message: 'Catégorie de maintenance invalide'
    },
    index: true
  },
  date: {
    type: Date,
    required: [true, 'La date est requise'],
    default: Date.now,
    index: true
  },
  kmAtService: {
    type: Number,
    required: [true, 'Le kilométrage au moment du service est requis'],
    min: [0, 'Le kilométrage ne peut pas être négatif']
  },
  cost: {
    type: Number,
    min: [0, 'Le coût ne peut pas être négatif'],
    default: 0
  },
  garageName: {
    type: String,
    trim: true,
    maxlength: [100, 'Le nom du garage ne peut pas dépasser 100 caractères']
  },
  invoiceFileUrl: {
    type: String,
    trim: true,
    maxlength: [500, 'L\'URL du fichier ne peut pas dépasser 500 caractères']
  },
  notes: {
    type: String,
    trim: true,
    maxlength: [1000, 'Les notes ne peuvent pas dépasser 1000 caractères']
  }
}, {
  timestamps: true
});

// Index composé pour les requêtes fréquentes
maintenanceLogSchema.index({ vehicleId: 1, date: -1 });
maintenanceLogSchema.index({ vehicleId: 1, category: 1 });
maintenanceLogSchema.index({ vehicleId: 1, date: -1, category: 1 });

// Virtuals pour compatibilité avec l'ancien code
maintenanceLogSchema.virtual('vehicle').get(function() {
  return this.vehicleId;
});

maintenanceLogSchema.virtual('vehicle').set(function(value) {
  this.vehicleId = value;
});

maintenanceLogSchema.virtual('type').get(function() {
  return this.category;
});

maintenanceLogSchema.virtual('type').set(function(value) {
  this.category = value;
});

maintenanceLogSchema.virtual('description').get(function() {
  return this.label;
});

maintenanceLogSchema.virtual('description').set(function(value) {
  this.label = value;
});

maintenanceLogSchema.virtual('km').get(function() {
  return this.kmAtService;
});

maintenanceLogSchema.virtual('km').set(function(value) {
  this.kmAtService = value;
});

maintenanceLogSchema.set('toJSON', { virtuals: true });
maintenanceLogSchema.set('toObject', { virtuals: true });

module.exports = mongoose.model('MaintenanceLog', maintenanceLogSchema);

File: flutter_app/lib/models/maintenance_log.dart
class MaintenanceLog {
  final String id;
  final String vehicleId;
  final String label;
  final String category;
  final DateTime date;
  final int kmAtService;
  final double cost;
  final String? garageName;
  final String? invoiceFileUrl;
  final String? notes;
  final DateTime createdAt;
  final DateTime updatedAt;

  MaintenanceLog({
    required this.id,
    required this.vehicleId,
    required this.label,
    required this.category,
    required this.date,
    required this.kmAtService,
    required this.cost,
    this.garageName,
    this.invoiceFileUrl,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
  });

  factory MaintenanceLog.fromJson(Map<String, dynamic> json) {
    return MaintenanceLog(
      id: json['_id'] ?? json['id'] ?? '',
      vehicleId: json['vehicleId'] ?? json['vehicle'] ?? '',
      label: json['label'] ?? json['description'] ?? '',
      category: json['category'] ?? json['type'] ?? '',
      date: json['date'] != null 
          ? DateTime.parse(json['date']) 
          : DateTime.now(),
      kmAtService: json['kmAtService'] ?? json['km'] ?? 0,
      cost: (json['cost'] ?? 0).toDouble(),
      garageName: json['garageName'],
      invoiceFileUrl: json['invoiceFileUrl'],
      notes: json['notes'],
      createdAt: json['createdAt'] != null 
          ? DateTime.parse(json['createdAt']) 
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null 
          ? DateTime.parse(json['updatedAt']) 
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'label': label,
      'category': category,
      'date': date.toIso8601String(),
      'kmAtService': kmAtService,
      'cost': cost,
      if (garageName != null && garageName!.isNotEmpty) 'garageName': garageName,
      if (invoiceFileUrl != null && invoiceFileUrl!.isNotEmpty) 'invoiceFileUrl': invoiceFileUrl,
      if (notes != null && notes!.isNotEmpty) 'notes': notes,
    };
  }
}



